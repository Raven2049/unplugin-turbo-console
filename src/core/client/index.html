<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>turbo console</title>
</head>
<style>
  #error-info {
    color: red;
    font-size: 2rem;
  }
</style>
<body>
  <div id="error-info"></div>
</body>
<script>
 (async () => {
   try {
     const hash = window.location.hash
     const encodeBase64 = window.atob(hash.slice(1))
     const base = new URL(window.location.href).searchParams.get('b') || '/'
     await fetch(`${base}__open-in-editor?file=${encodeURIComponent(encodeBase64)}`)
     window.close()
   }
   catch (error) {
     const errorInfo = document.getElementById('error-info')
     errorInfo.innerText = String(error)
   }
 })()
</script>
</html>
