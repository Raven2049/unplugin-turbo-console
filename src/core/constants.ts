export const PLUGIN_NAME = 'unplugin-turbo-console'
export const NUXT_CONFIG_KEY = 'turboConsole'
export const clientVirtualModuleId = `~console/client`
export const resolvedClientVirtualModuleId = `\0${clientVirtualModuleId}`
export const helperVirtualModuleId = `~console/helper`
export const resolvedHelperVirtualModuleId = `\0${helperVirtualModuleId}`
