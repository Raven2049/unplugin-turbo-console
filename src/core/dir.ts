import { dirname, resolve } from 'pathe'
import { fileURLToPath } from 'node:url'

export const DIR_DIST = typeof __dirname !== 'undefined'
  ? __dirname
  : dirname(fileURLToPath(import.meta.url))

export const DIR_CLIENT = resolve(DIR_DIST, './client/index.html')
