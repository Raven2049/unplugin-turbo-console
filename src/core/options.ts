import type { Options } from '../types'
import { env } from 'node:process'

const DEFAULT_OPTIONS: Options = {
  prefix: '',
  suffix: '',
  disableLaunchEditor: false,
  disableHighlight: false,
  disablePassLogs: false,
  extendedPathFileNames: [],
  babelParserPlugins: ['typescript', 'jsx'],
  port: 3070,
  silent: false,
}

export const BUILD_OPTIONS: Options = {
  disableLaunchEditor: true,
  disableHighlight: true,
  disablePassLogs: true,
}

export function resolveOptions(options: Options): Options {
  // merge babel parser plugins
  if (options.babelParserPlugins) {
    options.babelParserPlugins = [
      ...DEFAULT_OPTIONS.babelParserPlugins!,
      ...options.babelParserPlugins,
    ]
  }

  let resolved = {
    ...DEFAULT_OPTIONS,
    ...options,
  }

  if (env.NODE_ENV === 'production') {
    resolved = {
      ...resolved,
      ...BUILD_OPTIONS,
    }
  }

  return resolved
}
