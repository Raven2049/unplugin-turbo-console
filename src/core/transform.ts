import { basename, extname } from 'node:path'
import { ScriptTarget, createSourceFile, forEachChild, isCallExpression } from 'typescript'
import MagicString from 'magic-string'
import { parse as vueParser } from 'vue/compiler-sfc'
import type { Options, TraverseAST } from '../types'
import { getConsoleStyle } from './utils'

function traverseAST(args: TraverseAST) {
  const { node, s, id, options, offset = 0 } = args
  const fileName = basename(id)
  const fileType = extname(id)

  if (isCallExpression(node) as any) {
    if (node?.expression?.expression?.escapedText === 'console' && node?.expression?.name?.escapedText === 'log') {
      const argumentStart = node.arguments[0].pos + offset
      const argumentEnd = node.arguments[node.arguments.length - 1].end + offset
      const lineNumber = s.slice(0, argumentStart)?.split('\n').length

      const argsName = s.slice(argumentStart, argumentEnd).toString()
        .replaceAll('`', '')
        .replaceAll('\n', '')
        .replaceAll('\"', '')

      let { suffix, prefix } = options || {}
      suffix = suffix ? `\\n ${suffix}` : ''
      prefix = prefix ? `${prefix} \\n` : ''

      s.appendLeft(argumentStart, `"${prefix} %c${fileName}:${lineNumber} ~ ${argsName}","${getConsoleStyle(fileType)}",`)
      s.appendRight(argumentEnd, `,"${suffix}"`)
    }
  }

  forEachChild(node, childNode => traverseAST({
    node: childNode,
    s,
    id,
    options,
    offset,
  }))
}

export function transformer(code: string, id: string, options: Options) {
  if (['.js', '.ts', '.tsx', '.jsx'].includes(extname(id))) {
    const ast = createSourceFile(id, code, ScriptTarget.Latest)
    const s = new MagicString(code)
    traverseAST({
      node: ast,
      s,
      id,
      options,
    })
    return s.toString()
  }
  else if (extname(id) === '.vue') {
    const { descriptor } = vueParser(code, {
      sourceMap: true,
    })
    const s = new MagicString(code)
    if (descriptor.scriptSetup?.content) {
      const ast = createSourceFile(id, descriptor.scriptSetup?.content, ScriptTarget.Latest)
      traverseAST({
        node: ast,
        s,
        id,
        options,
        offset: descriptor.scriptSetup.loc.start.offset,
      })
    }

    if (descriptor.script?.content) {
      const ast = createSourceFile(id, descriptor.script?.content, ScriptTarget.Latest)
      traverseAST({
        node: ast,
        s,
        id,
        options,
        offset: descriptor.script.loc.start.offset,
      })
    }

    return s.toString()
  }
  return code
}
