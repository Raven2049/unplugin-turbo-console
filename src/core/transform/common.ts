import { relative } from 'node:path'
import { cwd } from 'node:process'
import { Buffer } from 'node:buffer'
import type { Node } from '@babel/types'
import type { GenContext } from '../../types'
import { getConsoleStyle, launchEditorStyle } from '../utils'

export function genConsoleString(genContext: GenContext) {
  const { options, fileName, originalColumn, originalLine, argType, filePath, argsName, fileType } = genContext
  const { prefix, suffix, disableLaunchEditor } = options

  const _prefix = prefix ? `${prefix} \\n` : ''
  const _suffix = suffix ? `\\n ${suffix}` : ''

  // not output when argtype is string or number
  const lineInfo = `${_prefix}%cðŸš€ ${fileName}:${originalLine}${['StringLiteral', 'NumericLiteral'].includes(argType) ? '' : ` ~ ${argsName}`}`
  const codePosition = `${relative(cwd(), filePath)}:${originalLine}:${(originalColumn || 0) + 1}`

  const launchEditorString = `%cðŸ”¦ Jump to Editor http://localhost:3000/client#${Buffer.from(codePosition, 'utf-8').toString('base64')}`

  let consoleString = ''

  if (!disableLaunchEditor)
    consoleString = `"${lineInfo} %c\\n${launchEditorString}","${getConsoleStyle(fileType)}","","${launchEditorStyle}","\\n",`

  else
    consoleString = `"${lineInfo} %c\\n","${getConsoleStyle(fileType)}","\\n",`

  return {
    consoleString,
    _suffix,
  }
}

export function isConsoleExpress(node: Node) {
  return node.type === 'CallExpression'
    && node.callee.type === 'MemberExpression'
    && node.callee.object.type === 'Identifier'
    && node.callee.object.name === 'console'
    && node.callee.property.type === 'Identifier'
    && node.callee.property.name === 'log'
    && node.arguments?.length > 0
}
