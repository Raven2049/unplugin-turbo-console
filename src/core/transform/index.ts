import type { Context } from './../../types'
import MagicString from 'magic-string'
import { parseSync } from 'oxc-parser'
import { genConsoleString, getCompiler, getLineAndColumn, isConsoleExpression, isPluginDisable } from '../utils'
import { walk } from '../utils/walker'
import { compilers } from './compilers'

export async function transform(context: Context) {
  const { code, id, options } = context
  const magicString = new MagicString(code)

  const compiler = await getCompiler(id)

  if (!compiler) {
    return {
      code: magicString.toString(),
      map: magicString.generateMap({
        source: id,
        file: id,
        includeContent: true,
        hires: true,
      }),
    }
  }

  const compileResult = await compilers[compiler](context)

  const ast = parseSync(id, compileResult.script, {
    lang: (compileResult.lang || 'js') as 'js' | 'jsx' | 'ts' | 'tsx' | undefined,
    sourceType: 'module',
  })

  if (isPluginDisable({
    comments: ast.comments || [],
    originalLine: 1,
    script: compileResult.script,
    id,
    type: 'top-file',
    compiler,
  })) {
    return {
      code: magicString.toString(),
      map: magicString.generateMap({
        source: id,
        file: id,
        includeContent: true,
        hires: true,
      }),
    }
  }

  walk(ast.program, {
    enter(node) {
      if (isConsoleExpression(node)) {
        const expressionStart = node.start
        const expressionEnd = node.end

        const originalExpression = magicString.slice(expressionStart, expressionEnd)

        if (originalExpression.includes('%c'))
          return false

        const { line, column } = getLineAndColumn(compileResult.script, expressionStart)

        const originalLine = line + compileResult.line
        const originalColumn = column

        if (isPluginDisable({
          comments: ast.comments || [],
          script: compileResult.script,
          originalLine: line,
          id,
          type: 'inline-file',
          compiler,
        })) {
          return false
        }

        // @ts-expect-error any
        const args = node.arguments

        const argsStart = args[0].start! + compileResult.offset
        const argsEnd = args[args.length - 1].end! + compileResult.offset
        const argType = args[0].type

        const argsName = magicString.slice(argsStart, argsEnd)
          .toString()
          .replace(/`/g, '')
          .replace(/\n/g, '')
          .replace(/"/g, '')

        const { consoleString, _suffix } = genConsoleString({
          options,
          originalLine,
          originalColumn,
          argType,
          argsName,
          id,
        })

        consoleString && magicString.appendLeft(argsStart, consoleString)
        _suffix && magicString.appendRight(argsEnd, `,"${_suffix}"`)
      }
    },
  })

  return {
    code: magicString.toString(),
    map: magicString.generateMap({
      source: id,
      file: id,
      includeContent: true,
      hires: true,
    }),
  }
}
