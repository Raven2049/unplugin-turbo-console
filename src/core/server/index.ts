import { createServer } from 'node:http'
import { readFile, stat } from 'node:fs/promises'
import { createApp, eventHandler, serveStatic, toNodeListener } from 'h3'
import { DIR_CLIENT } from './../dir'

export async function startServer() {
  try {
    await fetch(`http://localhost:${3000}/health`)
  }
  catch (error) {
    const app = createApp()
    app.use('/client', eventHandler(async (event) => {
      await serveStatic(event, {
        fallthrough: true,
        getContents: () => readFile(DIR_CLIENT),
        getMeta: async () => {
          return {
            type: 'text/html',
            encoding: 'utf8',
            size: (await stat(DIR_CLIENT)).size,
          }
        },
        indexNames: [`${DIR_CLIENT}/index.html`],
      })
    }))

    app.use('/health', eventHandler(() => {
      return {
        message: 'ok',
      }
    }))

    app.use('/__open-in-editor', eventHandler(async () => {
      return {
        body: 'ok',
      }
    }))
    createServer(toNodeListener(app)).listen(3000)
  }
}
