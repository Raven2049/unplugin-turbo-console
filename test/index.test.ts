import { describe, expect, it } from 'vitest'
import { transformer } from '../src/core/transform'
import { js, ts } from './fixtures/index'

describe('transformer', () => {
  it('transform js', () => {
    expect(transformer(js.code, js.id, js.options)).toMatchInlineSnapshot(`
      "
        const bar = 'bar'
        console.log(\\"🚀 \\\\n %cindex.js:3 ~ bar\\",\\"padding:4px; border-radius:5px; font-weight:600; color: #111827; background: #F7DF1E\\",bar,\\"\\\\n 🐶\\")
        "
    `)
  })

  it('transform ts', () => {
    expect(transformer(ts.code, ts.id, ts.options)).toMatchInlineSnapshot(`
      "
      const foo:string = 'foo'
      console.log(\\" %cindex.ts:3 ~ foo\\",\\"padding:4px; border-radius:5px; font-weight:600; color: #fff; background: #3178C6\\",foo,\\"\\")
        "
    `)
  })
})
