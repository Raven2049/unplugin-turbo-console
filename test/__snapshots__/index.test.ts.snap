// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`webpack js transform > script 1`] = `
{
  "code": "console.group(\\"%cðŸš€ main.js:1%cðŸ”¦ http://localhost:3090/client#Li4vLi4vbWFpbi5qczoxOjE=\\",\\"padding:2px 5px; border-radius:3px 0 0 3px;margin-bottom:5px;color: #111827; background: #F7DF1E\\",\\"background: #00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin-bottom:5px\\")
console.log('hello javascript')
console.groupEnd()",
  "map": SourceMap {
    "file": undefined,
    "mappings": ";AAAA",
    "names": [],
    "sources": [
      "main.js",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`webpack vue transform > script 1`] = `
{
  "code": "<template>
  <img alt=\\"Vue logo\\" src=\\"./assets/logo.png\\">
  <HelloWorld msg=\\"Welcome to Your Vue.js + TypeScript App\\"/>
</template>

<script lang=\\"ts\\">
import { defineComponent } from 'vue';
import HelloWorld from './components/HelloWorld.vue';

console.group(\\"%cðŸš€ App.vue:10%cðŸ”¦ http://localhost:3070/client#Li4vLi4vQXBwLnZ1ZToxMDox\\",\\"padding:2px 5px; border-radius:3px 0 0 3px;margin-bottom:5px;color: #fff; background: #4FC08D\\",\\"background: #00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin-bottom:5px\\")
console.log('App.vue')
console.groupEnd()

export default defineComponent({
  name: 'App',
  components: {
    HelloWorld
  }
});
</script>
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;;AACA;kBAAsB;AACtB;AACA;AACA;AACA;AACA;AACA;AACA;AACA;",
    "names": [],
    "sources": [
      "App.vue",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`webpack vue transform > script setup 1`] = `
{
  "code": "<script setup lang=\\"ts\\">
  console.group(\\"%cðŸš€ App.vue:2%cðŸ”¦ http://localhost:3090/client#Li4vLi4vQXBwLnZ1ZToyOjM=\\",\\"padding:2px 5px; border-radius:3px 0 0 3px;margin-bottom:5px;color: #fff; background: #4FC08D\\",\\"background: #00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin-bottom:5px\\")
console.log('script setup')
console.groupEnd()
  </script>
  
  <template>
    <div>
      script setup
    </div>
  </template>
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AAAE;kBAA2B;AAC7B;AACA;AACA;AACA;AACA;AACA;AACA;",
    "names": [],
    "sources": [
      "App.vue",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;
